### PR TITLE
feat(TCOMP-2107): Implement a stop strategy for streaming input connectors

### PR DESCRIPTION
**Pull request by undx**
https://jira.talendforge.org/browse/TCOMP-2107

### CODE REVIEW

File: component-runtime-impl/src/main/java/org/talend/sdk/component/runtime/input/Streaming.java [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+                            .map(Long::parseLong)
+                            .orElse(500L));
+        }
+    }
+
+    public static StopStrategy loadStopStrategy(final String plugin) {
+        final LocalConfiguration configuration = ofNullable(ContainerFinder.Instance.get().find(plugin))
+                .map(it -> it.findService(LocalConfiguration.class))
+                .orElseGet(defaultLocalConfiguration);
+        final Long maxReadRecords = ofNullable(configuration.get("talend.input.streaming.maxRecords"))
```
Comment (ozhelezniak-talend) [inline component-runtime-impl/src/main/java/org/talend/sdk/component/runtime/input/Streaming.java:91]: Can I configure those configuration options in the component UI?
talend.input.streaming.maxRecords
talend.input.streaming.maxDurationMs

**Review by ozhelezniak-talend — COMMENTED** [review]
We've lost an ability to endlessly run the component. If the max record and max timeout isn't set the component won't run at all.
Is it expected?

Do we guarantee that those two new options can't be equal -1?

File: component-runtime-impl/src/main/java/org/talend/sdk/component/runtime/input/Streaming.java [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+
+        public StopConfiguration(final Long maxRecords, final Long maxTime, final Long start) {
+            maxReadRecords = maxRecords == null ? -1L : maxRecords;
+            maxActiveTime = maxTime == null ? -1L : maxTime;
+            started = start == null ? System.currentTimeMillis() : start;
+        }
+
+        @Override
+        public boolean isActive() {
+            return (maxReadRecords > -1) || (maxActiveTime > -1);
```
[the remark this answers is not in the document]
Comment (undx) [reply component-runtime-impl/src/main/java/org/talend/sdk/component/runtime/input/Streaming.java:204]: By default, the stop stragegy conditions are both set to -1, this make the stragegy inactive. When it's inactive, the job runs endlessly. Only the retry stragegy can stop it!

File: component-runtime-impl/src/main/java/org/talend/sdk/component/runtime/input/Streaming.java (hunk 1 above)
Comment (undx) [reply component-runtime-impl/src/main/java/org/talend/sdk/component/runtime/input/Streaming.java:91]: By default, first element checked is localconfiguration then system property. 
We can have a mapping with localconf on UI like this 
```
@Option
@DefaultValue("local_configuration:myfamily.model.key")
private String value;
```
As stated by documentation, some action need to be done to propagate information.

File: component-runtime-impl/src/main/java/org/talend/sdk/component/runtime/input/Streaming.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (ozhelezniak-talend) [reply component-runtime-impl/src/main/java/org/talend/sdk/component/runtime/input/Streaming.java:204]: yes, you totally right. I misread the condition where the isActive is used.

File: component-runtime-impl/src/main/java/org/talend/sdk/component/runtime/input/Streaming.java (hunk 1 above)
Comment (ozhelezniak-talend) [reply component-runtime-impl/src/main/java/org/talend/sdk/component/runtime/input/Streaming.java:91]: yes, in this way we can have this value in UI. 
But as a user, I want to pass the value from the UI to the configuration to stop watch.

**Review by ozhelezniak-talend — APPROVED** [review]
LGTM
Great feature!

File: component-runtime-impl/src/main/java/org/talend/sdk/component/runtime/input/Streaming.java (hunk 1 above)
Comment (undx) [reply component-runtime-impl/src/main/java/org/talend/sdk/component/runtime/input/Streaming.java]: https://jira.talendforge.org/browse/TCOMP-2226

**Pull request merged**